### PR TITLE
Update Rust and Haskell versions.

### PR DESCRIPTION
**Pull request by td202**
## Purpose

Update the Rust version to 1.82 and the GHC version to 9.6.6 (LTS-22.39).

## Changes

- Update versions.
- Update base image tag to `rust-1.82_ghc-9.6.6-1`.
- Clippy fixes.
- Update rust library versions.

## Checklist

- [X] My code follows the style of this project.
- [X] The code compiles without warnings.
- [X] I have performed a self-review of the changes.
- [X] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [X] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: rust-src/concordium_base/src/id/account_holder.rs [hunk 1]
```diff
@@ -704,11 +701,7 @@ where
 }
 
 #[allow(clippy::too_many_arguments)]
-pub fn create_unsigned_credential<
-    P: Pairing,
-    C: Curve<Scalar = P::ScalarField>,
-    AttributeType: Attribute<C::Scalar>,
->(
+pub fn create_unsigned_credential<P: Pairing, C: Curve<Scalar = P::ScalarField>, AttributeType>(
```
Comment (soerenbf) [inline rust-src/concordium_base/src/id/account_holder.rs:704]: **Non-blocking**: Is this clippy? I'm wondering why the trait bound declarations are not consistent 🤷

**Review by soerenbf — APPROVED** [review]
LGTM! Good job going through all this 😅 

CI fails though so there's that 🤷

File: .github/workflows/build-test-sources.yaml [hunk 2]
```diff
@@ -85,7 +85,7 @@ jobs:
     strategy:
       matrix:
         plan:
-        - rust: 1.73
+        - rust: 1.82
```
Comment (soerenbf) [inline .github/workflows/build-test-sources.yaml:88]: I'm wondering, is it necessary to test that we can build all this with the MSRV as well? Or is it enough that it is specified in `cargo.toml`?

File: rust-src/concordium_base/src/id/account_holder.rs (hunk 1 above)
Comment (td202) [reply rust-src/concordium_base/src/id/account_holder.rs]: Clippy doesn't like having bounds in both the parameter and the where clause, and suggested having them in the where clause. But I think it's probably better to have them in the parameter.

File: .github/workflows/build-test-sources.yaml (hunk 2 above)
Comment (td202) [reply .github/workflows/build-test-sources.yaml:88]: You're right. The `check-base-compiles` job is supposed to do this, so I'll set that to the MSRV.

**Pull request merged**